ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/BlockResultFactory.java

```diff
@@ -95,19 +99,29 @@ public EngineGetPayloadResultV1 payloadTransactionCompleteV1(final Block block)
     return new EngineGetPayloadResultV1(block.getHeader(), txs);
   }
 
-  public EngineGetPayloadResultV2 payloadTransactionCompleteV2(final Block block) {
+  public EngineGetPayloadResultV2 payloadTransactionCompleteV2(
+      final BlockWithReceipts blockWithReceipts) {
     final List<String> txs =
-        block.getBody().getTransactions().stream()
+        blockWithReceipts.getBlock().getBody().getTransactions().stream()
             .map(TransactionEncoder::encodeOpaqueBytes)
             .map(Bytes::toHexString)
             .collect(Collectors.toList());
 
-    final long blockValue = calculateBlockValue(txs);
-    return new EngineGetPayloadResultV2(block.getHeader(), txs, Quantity.create(blockValue));
+    final long blockValue = calculateBlockValue(blockWithReceipts);
+    return new EngineGetPayloadResultV2(
+        blockWithReceipts.getHeader(), txs, Quantity.create(blockValue));
   }
 
-  private long calculateBlockValue(final List<String> ignored) {
-    return 0L;
+  private long calculateBlockValue(final BlockWithReceipts blockWithReceipts) {
+    final Block block = blockWithReceipts.getBlock();
+    final List<Transaction> txs = block.getBody().getTransactions();
+    final List<TransactionReceipt> receipts = blockWithReceipts.getReceipts();
+    Wei totalFee = Wei.of(0);
+    for (int i = 0; i < txs.size(); i++) {
```

Comment:
Using a for each would be more idiomatic.

```
for (Transaction tx : txs) {
...
}
```

Reply:
Need to traverse through transactions as well as the receipts, could potentially use two iterators and a while loop to do this, what do you think?

Reply:
Didn't spot the receipts access as well, what you have is good then.